components/butterfly/src/rumor.rs

```diff
@@ -326,41 +372,26 @@ mod storage {
         }
     }
 
-    impl<T> RumorStore<T> where T: Rumor
-    {
-        /// Create a new RumorStore for the given type. Allows you to initialize the counter to a
-        /// pre-set value. Useful mainly in testing.
-        pub fn new(counter: usize) -> RumorStore<T> {
-            RumorStore { update_counter: Arc::new(AtomicUsize::new(counter)),
-                         ..Default::default() }
-        }
-
-        /// Clear all rumors and reset update counter of RumorStore.
-        pub fn clear(&self) -> usize {
-            let mut list = self.list.write();
-            list.clear();
-            self.update_counter.swap(0, Ordering::Relaxed)
-        }
-
-        pub fn encode(&self, key: &str, member_id: &str) -> Result<Vec<u8>> {
-            let list = self.list.read();
-            match list.get(key).and_then(|l| l.get(member_id)) {
-                Some(rumor) => rumor.clone().write_to_bytes(),
-                None => Err(Error::NonExistentRumor(String::from(member_id), String::from(key))),
-            }
+    impl<T: Rumor> RumorStore<T> {
+        /// # Locking
+        /// * `RumorStore::list` (read) This method must not be called while any RumorStore::list
+        ///   lock is held.
+        pub fn encode_rsr(&self, key: &str, member_id: &str) -> Result<Vec<u8>> {
+            self.map_rumor_rsr(key, member_id, T::write_to_bytes)
+                .unwrap_or_else(|| {
+                    Err(Error::NonExistentRumor(String::from(member_id), String::from(key)))
+                })
```

Comment:
This is a great simplification that I can take advantage of in my rumor work as well (I implemented something very similar).